### PR TITLE
Llama 3 Support

### PR DESCRIPTION
**Pull request by Rexwang8**
This PR aims to update gpt_bpe to support uint32 and Llama 3. The change from uint16 to uint32 is in order to allow gpt_bpe to support vocab sizes greater than 65535 in length like Llama 3. Effort has been made to both clean up the repo and update it to go1.22, passing a number of previously broken tests.

## Changes

1. Replaced uint16 internal representation with uint32, as a prerequisite to support larger vocab sizes.
2. Added the data and embedded Llama 3. Data was sourced from Meta's huggingface repo. (https://huggingface.co/meta-llama/Meta-Llama-3-8B). 
3. Added a `getVocab` function as part of the resolver so we don't need to re-read the vocab multiple times through each encoding.
4. Updated default pad token handling to use uint16 if vocab size is not uint32, and uint32 if it is.
5. Added handling for the special tokens subset format found in `tokenizer.json` in llama 3 models, which list special tokens but excludes them from the vocab. Added functionality to resolve and re-add these tokens if found.
6. Added additional handling to get merges from a `tokenizer.json` file and added a `getMerges` function similar to # 3 above
7. Added proper reading and handling of TokenizerClass, IgnoreMerges and NewLineMode from `tokenizer.json` and `config.json`
8. Added reading and handling of splitRegex from `tokenizer.json`
9. Update repo to go1.22 and clean up various deprecated functions.

## Tests
1. All previously working tests still pass
2. Added encode, decode, frankenstein, encode and decoding, padding tests for Llama 3
11. Added download and remote tokenization tests for Llama 3 and Mistral
12. Cleaned up the format of download tests somewhat with the use of a helper
13. Cleaned up parts of the test file (`gpt_bpe_test.go`)
14. Added tests to dataset_tokenizer.go to ensure encode and decode works properly
15. Added binary large file encode/decode tests for Llama 3 and Mistralv1, which includes a number of edge cases.
16. Added a few misc test cases covering previously encountered edge cases.
17. Re-added previously broken nerdstash test reference file.
18. Added some commenting to help people understand the purpose of the tests.
19. Updated tests for `dataset_tokenizer.go`

## Notable bugs fixed
1.  Resolved the check with merges in streaming encode, the `ok` variable was always true even if the merge was not found in the dictonary, defaulting to a value of `0`, which could cause issues in edge cases.
2. Resolved sentancepiece converter as a part of the model download process not adding split files to `resolvedResources` after splitting.
3. Added multi-token support for byte tokens (ie Mistral Emojis)
4. Prevent the tokenizer from splitting special tokens that consist of multiple special tokens like the NerdStash's space `<space><space><space><space>` tokens
5. Added an exception for `ignoreMerges` configured models like llama3 and mistral-nemo which skips accumulator token merging in order to handle some edge cases like two words merging into one.
6. Resolved `sampling` for `dataset_tokenizer` to be able to use gcd to choose how often to skip contexts.
7. Fixed `shuffling` for `dataset_tokenizer` when writing contexts creating junk data for smaller than full contexts.

### CODE REVIEW

File: gpt_bpe_test.go [hunk 1]
```diff
@@ -907,80 +1006,112 @@ func TestPythiaRemoteDownloadTokenizer(t *testing.T) {
...
 }
 
-func TestModelDownload(t *testing.T) {
+func downloadModel(modelId string, destPath string) error {
 	// Download the model
-	modelId := "gpt2"
-	destPath := "./TestModelDownload"
 	destPathPTR := &destPath
-
 	rsrcType, hfApiToken := resources.RESOURCETYPE_TRANSFORMERS, os.Getenv("HF_API_TOKEN")
```
Comment (harubaru) [inline gpt_bpe_test.go:1065]: This might be out of scope so I'm just bringing attention to it, HF updated the token env var name to `HF_TOKEN`.

File: gpt_bpe.go [hunk 2]
```diff
@@ -1125,14 +1225,18 @@ func (encoder *GPTEncoder) EncodeBuffer(buffer *[]byte) *[]byte {
...
-		binary.Write(buf, binary.LittleEndian, tokens)
+		_ = binary.Write(buf, binary.LittleEndian, tokens)
 	}
 	bufBytes := buf.Bytes()
 	return &bufBytes
 }
 
 // Encode encodes a string into a sequence of tokens.
 func (encoder *GPTEncoder) Encode(text *string) *Tokens {
+	// Temporary hack - inject a space token at the end of the accumulator for mistral-tokenizer
```
Comment (rtalaricw) [inline gpt_bpe.go:1236]: I assume this where the pre-processing, normalization and post-processing would happen?!

Comment (wbrown) [reply gpt_bpe.go:1236]: Yes, exactly.

File: gpt_bpe_test.go [hunk 3]
```diff
@@ -1193,141 +1532,85 @@ func TestModelDownloadLlama(t *testing.T) {
...
-	if _, err := os.Stat(tokenizerConfigPath); err == nil {
-		fmt.Println("tokenizer.model exists")
-
-	} else if errors.Is(err, os.ErrNotExist) {
-		t.Errorf("tokenizer.model does not exist. %s", err)
-
-	} else {
-		t.Errorf("Error checking for tokenizer.model. %s", err)
-	}
+	assertFileExists(t, tokenizerConfigPath)
```
Comment (rtalaricw) [inline gpt_bpe_test.go:1535]: I like how keeping this separate provides readability but this can be condensed into a for loop

File: gpt_bpe_test.go [hunk 4]
```diff
@@ -1193,141 +1532,85 @@ func TestModelDownloadLlama(t *testing.T) {
...
-	if _, err := os.Stat(modelPath); err == nil {
-		fmt.Println("pytorch_model.bin exists")
-
-	} else if errors.Is(err, os.ErrNotExist) {
-		t.Errorf("pytorch_model.bin does not exist")
-
-	} else {
-		t.Errorf("Error checking for pytorch_model.bin")
-	}
+	assertFileExists(t, modelPath)
```
Comment (rtalaricw) [inline gpt_bpe_test.go:1605]: Same here

**Review by rtalaricw — APPROVED** [review]
I think this PR actually marks the "corner" for us in terms of getting the finetuner, tokenizer itself (this), and inference in really good shape.

File: resources/data/llama3-tokenizer/config.json [hunk 5]
```diff
@@ -0,0 +1,27 @@
+{
```
Comment (harubaru) [inline resources/data/llama3-tokenizer/config.json:1]: Does this need to be here along with `resources/data/llama3-tokenizer/generation_config.json`?

**Review by harubaru — APPROVED** [review]
Even as I went through everything, I don't have any other feedback to add on top of what Rahul had said, other than some nitpicking. The refactor is a welcome touch too. I also really like the new comments especially that in `ToBPE()`, as they help explain the behavior of what's going on especially for ourselves in the future if we need to come back to change something.

File: resources/data/llama3-tokenizer/config.json (hunk 5 above)
Comment (Rexwang8) [reply resources/data/llama3-tokenizer/config.json:1]: you're right, we don't actually use generation_config when resolving a tokenizer, we should remove this.

File: cmd/dataset_tokenizer/dataset_tokenizer.go [hunk 6]
```diff
@@ -1106,8 +1137,26 @@ func WriteContexts(outPath string, contexts chan gpt_bpe.Tokens,
...
+				// Move the paddingTuple to new position if necessary
+				for i, idx := range idxes {
+					if idx.contextStart == int(prevTarget) && idx.contextEnd == int(prevTarget)+contextSize {
+						// Move to end of file
+						idxes[i].contextStart = endpos
+						idxes[i].contextEnd = endpos + contextSize
+					}
+				}
+
+				idxes = append(idxes, paddingTuple{start: lenBeforePad, end: contextSize, contextStart: int(target), contextEnd: int(target) + contextSize})
```
Comment (wbrown) [inline cmd/dataset_tokenizer/dataset_tokenizer.go:1155]: Wrap this to 78.

File: cmd/dataset_tokenizer/dataset_tokenizer.go [hunk 7]
```diff
@@ -1018,6 +1019,16 @@ func (tt TextsTokenizer) TokenizeTextsToContexts(
 	return contexts, nil
 }
 
+// Greatest common divisor (GCD) via Euclidean algorithm
```
Comment (wbrown) [inline cmd/dataset_tokenizer/dataset_tokenizer.go:1022]: `// GCD ... ` is the proper function comment signature.

File: cmd/dataset_tokenizer/dataset_tokenizer.go [hunk 8]
```diff
@@ -1106,8 +1137,26 @@ func WriteContexts(outPath string, contexts chan gpt_bpe.Tokens,
...
 		if endpos > 0 && shuffle {
+			// Pad to context size if necessary
+			if len(*binContext) < contextSize {
+				lenBeforePad := len(*binContext)
+				for i := len(*binContext); i < contextSize; i++ {
+					(*binContext) = append(*binContext, byte(0))
+				}
+				// Move the paddingTuple to new position if necessary
+				for i, idx := range idxes {
+					if idx.contextStart == int(prevTarget) && idx.contextEnd == int(prevTarget)+contextSize {
```
Comment (wbrown) [inline cmd/dataset_tokenizer/dataset_tokenizer.go:1148]: Wrap.

File: cmd/dataset_tokenizer/dataset_tokenizer.go [hunk 9]
```diff
@@ -1132,6 +1181,57 @@ func WriteContexts(outPath string, contexts chan gpt_bpe.Tokens,
...
+			if err != nil {
+				if err == io.EOF {
+					break
+				}
+				return totalTokens, err
+			}
+			if num == 0 {
+				break
+			}
+			thisContext := paddingTuple{contextStart: currentStartPos, contextEnd: currentEndPos, start: 0, end: len(buf)}
```
Comment (wbrown) [inline cmd/dataset_tokenizer/dataset_tokenizer.go:1208]: Wrap

File: cmd/dataset_tokenizer/dataset_tokenizer.go [hunk 10]
```diff
@@ -1132,6 +1181,57 @@ func WriteContexts(outPath string, contexts chan gpt_bpe.Tokens,
...
+			// Check if the context is in the idxes
+			for _, idx := range idxes {
+				if idx.contextStart == thisContext.contextStart && idx.contextEnd == thisContext.contextEnd {
```
Comment (wbrown) [inline cmd/dataset_tokenizer/dataset_tokenizer.go:1211]: Wrap.

**Review by wbrown — CHANGES_REQUESTED** [review]
You should re-wrap `dataset_tokenizer.go`.

**Review by wbrown — APPROVED** [review]
